client/state/plugins/marketplace/actions.ts

```diff
@@ -40,3 +67,168 @@ export function setIsPluginInstalledDuringPurchase(
+					case PLUGIN_INSTALLATION_COMPLETED:
+						dispatch( pluginInstallationStateChange( MARKETPLACE_ASYNC_PROCESS_STATUS.COMPLETED ) );
+						break;
+					case PLUGIN_INSTALLATION_IN_PROGRESS:
+						dispatch(
+							pluginInstallationStateChange( MARKETPLACE_ASYNC_PROCESS_STATUS.IN_PROGRESS )
+						);
+						break;
+					case PLUGIN_INSTALLATION_ERROR:
+						dispatch(
+							pluginInstallationStateChange(
+								MARKETPLACE_ASYNC_PROCESS_STATUS.ERROR,
+								'Error in plugin installation in the backend'
+							)
+						);
+						break;
+				}
+			}
+		}
+	};
+}
+
+export function trySiteTransfer( selectedSiteId: number, pluginSlugToBeInstalled: string ) {
+	return function ( dispatch: any, getState: () => any ): void {
+		const state = getState();
+		const { siteTransferStatus } = getPurchaseFlowState( state );
+		// First, figure out if the site is already transferred
+		const { fetchingStatus: isFetchingTransferState, status } = getAutomatedTransfer(
+			state,
+			selectedSiteId
+		);
+
+		if ( siteTransferStatus === MARKETPLACE_ASYNC_PROCESS_STATUS.UNKNOWN ) {
+			dispatch( fetchAutomatedTransferStatus( selectedSiteId ) );
+			dispatch( siteTransferStateChange( MARKETPLACE_ASYNC_PROCESS_STATUS.FETCHING ) );
+		} else if ( ! isFetchingTransferState ) {
+			switch ( status ) {
+				case transferStates.COMPLETE:
+					dispatch( siteTransferStateChange( MARKETPLACE_ASYNC_PROCESS_STATUS.COMPLETED ) );
+					break;
+				case transferStates.NONE:
+					// The transferStates.NONE state corresponds to when the transfer details request returns the error response with message "An invalid transfer ID was passed.""
+					// This actually means there is no transfer that exists in the current site
+					dispatch( initiateThemeTransfer( selectedSiteId, null, pluginSlugToBeInstalled ) );
+					dispatch( siteTransferStateChange( MARKETPLACE_ASYNC_PROCESS_STATUS.IN_PROGRESS ) );
+					break;
+				case transferStates.ERROR:
+					dispatch(
+						siteTransferStateChange(
+							MARKETPLACE_ASYNC_PROCESS_STATUS.ERROR,
+							'The fetched backend transfer status indicated an error state'
+						)
+					);
+			}
+		}
+	};
+}
+
+/**
+ * This action is meant to be triggered multiple time times based on changes to states in async processes
+ * TODO: Write a comprehensive unit test for this function
+ */
+export function tryProductInstall() {
+	return function ( dispatch: any, getState: () => any ): void {
+		const state = getState();
+		const selectedSiteId = getSelectedSiteId( state );
+		const { pluginSlugToBeInstalled } = getPurchaseFlowState( state );
+		const { pluginInstallationStatus, siteTransferStatus } = getPurchaseFlowState( state );
+
+		if (
+			selectedSiteId &&
+			pluginSlugToBeInstalled &&
+			pluginInstallationStatus !== MARKETPLACE_ASYNC_PROCESS_STATUS.ERROR &&
+			siteTransferStatus !== MARKETPLACE_ASYNC_PROCESS_STATUS.ERROR
+		) {
+			if ( siteTransferStatus !== MARKETPLACE_ASYNC_PROCESS_STATUS.COMPLETED ) {
```

Comment:
Noting that these improvised state machines can be a bit difficult to debug.

In my experience, I've found that async thunks are much more straightforward to work with. Have you considered using them for this project?

Reply:
Since this isn't a straightforward waiting of several async actions, I am not clear on how to use async thunks here. Can you elaborate a bit more?

Reply:
Sure thing. 

To explain the problem: it appears like `tryPluginInstall` for example handles a bunch of states. This means `tryPluginInstall` has to be called multiple times in the meantime, during the time when those different states are being updated. That's all fine, however it can be fragile and difficult to diagnose and debug, because of potentially unpredictable ways state can be changed in the meantime, and unclear/undefined chain of events that the flow consists of. I'm happy to be convinced otherwise, and that's easy with some unit tests, however, I don't see any in this PR.

From a semantic point of view, calling `tryPluginInstall` does feel odd too - it's weird to call `tryPluginInstall` multiple times until all dependencies are resolved. I'd expect a single `tryPluginInstall` call, after all dependencies are resolved. 

Essentially, this can be achieved by implementing async thunks that await on promises that resolve when the dependency data is being resolved.

Anyway, I believe this would require a complete refactor of the logic, which I don't want to force you to do. I just wanted to offer another perspective and a solution that can be more predictable, straightforward and easy to follow when looking at the code.

Reply:
@tyxla thanks for raising this. I completely agree with all of you said. I am just a bit not clear how to queue and wait for a set of pre existing actions that already exist in the codebase that do not return any promises. Do you mean to just manually trigger the async api calls without using existing redux async actions? This would definitely make things more predictable. However as you said would require a rewrite. Please do tell me if I am on the right path!

Also @southp what do you think, do we have enough time to look at this with a different approach, dig in to level API calls without relying on existing actions?

Reply:
Normally I have worked with chaining thunks like this. https://github.com/reduxjs/redux/issues/1676#issuecomment-215413478

I am not fully aware of how we do it in calypso's middleware though, so sorry if I am treading down the wrong path 😿

Reply:
For example an action call like this returns a plain js object 

https://github.com/Automattic/wp-calypso/blob/f9ba00294f65ee421e36c3241338b48bf33eefa7/client/state/automated-transfer/actions.js#L35-L44

Reply:
@jdc91 if you search for `async ( dispatch` in the codebase, you'll find plenty of async thunks that rely on `async/await`. Those are all examples of async thunks that we've recently been adopting together with the migration to `react-query` and as we've been removing unnecessary data layer, redux and `wpcom-undocumented` instances.

I understand your concern regarding data layer handlers that take care of the API requests - to get around this, we usually refactor away from the data layer, and perform the `wpcom.get`/`wpcom.post` (or any other means of API requests) directly in the thunk. That allows us to `await` for the call, as well as to implement proper error handling directly in place. You'll notice that those action thunks are very easy to understand: it's immediately clear what they are doing while reading them; one doesn't need to dig into 10 more files to understand the full flow. That's why they're preferred - they're not only simple and concise; they're much easier to maintain and understand from a developer who's unfamiliar with them.

Let me know if that helps or if I can help any further.

Reply:
So just to be clear is it ok to just do the direct relevant API calls directly from the marketplace/actions.ts files? 
Or should the existing Actions be refactored and then used here?

Reply:
Yes, well both are fine - both performing the API calls directly in the action thunks, and dispatching other thunks that perform the API calls. The latter can also be async and `await`ed.

Reply:
Thanks! Since this PR has grown a bit, I will be looking at this in a follow up PR.